### ISSUE TITLE
How can I set up auto-increment/auto-timestamp columns?

### ISSUE THREAD

**Issue by csterritt** [citycide/trilogy#5]
Hey - trilogy looks great, but I'd like to have an 'id' column which auto-increments for each added row, and automatically set 'timestamps'.

I see that knex supports them at table creation:

``` js
knex.schema.createTable('users', function (table) {
  table.increments();
  table.string('name');
  table.timestamps();
})
Outputs:
create table `users` (`id` int unsigned not null auto_increment primary key, `name` varchar(255), `created_at` datetime, `updated_at` datetime)
```

But I don't see how to get that functionality with trilogy.

Thanks!

**Comment by citycide** [comment]
Hmm. This isn't something I've supported first-class yet but it is something I think should be there. In the meantime you can pull the knex schema builder and build it the way you're used to. Something like:

```javascript
// shorter version is obviously `db.getSchemaBuilder().createTable(...)`
const sb = db.getSchemaBuilder();
const query = sb.createTable('users', function (table) {
  table.increments();
  table.string('name');
  table.timestamps();
});

// now either pass as-is to `run()`
db.run(query)
  .then(function () {
    console.log('Done!');
  })
  .catch(function (e) {
    console.log('Uh oh!');
  });

// or make it a string explicitly, then pass it to `raw()`
db.raw(query.toString()).then(...);
```

I think it's actually possible right now, but it's not obvious and still isn't great once you know about it. You could theoretically pass it like this:
```javascript
db.createTable('users', [
  'name',
  { name: 'id', type: 'increments' },             // ew
  { name: 'unnecessary', type: 'timestamps' }     // gross
]);
```

If I were to add this, would it make sense to you for it to be in the `options` object passed as the 3rd arg to `createTable()`? ie.

```javascript
db.createTable('users', ['name'], { increments: true, timestamps: true });

// or maybe an array of `schema` props within the options object would make more sense:
db.createTable('users', ['name'], { schema: ['increments', 'timestamps'] });
```

The second possibility wouldn't allow for things that take arguments.

Please excuse any code errors above, it's example fluff :smiley:

**Comment by csterritt** [comment]
Hey @citycide thanks for the swift reply!  I'll try those workarounds out.

And yeah, the proposals look good -- maybe start with the first one, and go to (some variant of) the second one if we need more kinds of customizations?  For now simpler might be best...

**Comment by csterritt** [comment]
I tried the 'ew/gross' :-) option, and it worked fine.

However, when I insert stuff, it does indeed do the auto_increment id, but it puts nothing in for created_at/updated_at.

I guess I'm used to Rails' ActiveRecord giving values for these fields automagically -- should this be done here?  I guess knex doesn't do this?

**Comment by citycide** [comment]
Hmm - these are the Knex docs for `timestamps()`:

> timestamps — table.timestamps()
Adds a created_at and updated_at column on the database, setting these each to dateTime types. When true is passed as the first argument a timestamp type is used. Both colums default to being not null and the current timestamp when true is passed as the second argument.

You might have to try the raw way for now since there's not a way in trilogy to pass the arguments you'll need to `timestamps()`. It looks like you'd do something similar to `table.timestamps(true, true)`. I'm not sure since this isn't a feature I've ever used. I'm short on time lately but I'll see if I can test it out, though I'd appreciate any testing you can do yourself also.

For `updated_at` specifically - I don't know how Knex handles that one. It might include an `on update` clause in the query so sqlite handles it. Try to get the query Knex builds for you using either verbose mode ( see [here](https://citycide.github.io/trilogy/Trilogy.html) ) or the `toString()` method like I showed above.

**Comment by csterritt** [comment]
Thanks, I'll try that out.

**Comment by citycide** [comment]
@csterritt not sure if you've had the time to look into what Knex produces, but I put together a few quick queries as rough tests.


#### 1. Query:
```javascript
knex.schema.createTable('users', table => {
  table.increments()
  table.string('name')
  table.timestamps()
})
```

#### 1. Outputs:
```sql
create table "users" (
  "id" integer not null primary key autoincrement,
  "name" varchar(255),
  "created_at" datetime,
  "updated_at" datetime
)
```

#### 2. Query:
```javascript
knex.schema.createTable('users', table => {
  table.increments()
  table.string('name')
  table.timestamps(true)
})
```

#### 2. Outputs:
This outputs the same SQL string as 1, so the first argument seems to do nothing ( at least by itself, see query 3 ). This is probably due to differences in dialect support among sqlite, pg, mysql, etc.

#### 3. Query:
```javascript
knex.schema.createTable('users', table => {
  table.increments()
  table.string('name')
  table.timestamps(true, true)
})
```

#### 3. Outputs:
```sql
create table "users" (
  "id" integer not null primary key autoincrement,
  "name" varchar(255),
  "created_at" datetime not null default CURRENT_TIMESTAMP,
  "updated_at" datetime not null default CURRENT_TIMESTAMP
)
```

---

So based on that, using `timestamps(true, true)` would get you close. Anything you insert would be given  the current time in both date columns. As far as I can tell, `updated_at` would need to be changed manually using `CURRENT_TIMESTAMP` or `DATETIME('now')` within any `update`s you do on a row.

Hope this helps. I'll keep looking into adding some sugar for methods like `increments` and `timestamps` like we went into above.

**Comment by csterritt** [comment]
@citycide Thanks for the excellent help!  I've been dealing with the UI side of my current project, and that's pretty much done, so it's time to get back into the data side.  So I really appreciate your timeliness!  I'll give it a go; I spent some time with the knex documentation, and I think I just need to dig into it.

Thanks again!

**Comment by citycide** [comment]
@csterritt How has your usage been? I'm looking at making some changes ( I think all backwards-compatible ) to the table creation. Specifically there will be two additional methods of putting together your schema.

There's the current (Array) option:

```javascript
db.createTable('users', [
  'name', // defaults to text / string type
  { name: 'age', type: 'integer' },
  { name: 'favorite_number': type: 'integer', defaultTo: 6 },
  { name: 'likes_shrimp', defaultTo: false }
])
```

Now there will be Object syntax:

```javascript
db.createTable('users', {
  name: { type: 'text' },
  age: { type: 'integer' },
  favorite_number: { type: 'integer', defaultTo: 6 },
  likes_shrimp: { defaultTo: false }
})
```
And also Function syntax:

```javascript
let userSchema = function (table) {
  table.text('name')
  table.integer('age')
  table.integer('favorite_number').defaultTo(6)
  table.boolean('likes_shrimp').defaultTo(false) // is actually text / string in sqlite
}

db.createTable('users', userSchema)
```

While I'm still trying to think of a way to fit some of knex's methods like `timestamps` into the API, I think the Function syntax would become the de facto version for when you need to use these. Considering your use case and our above discussion, what do you think?

**Comment by csterritt** [comment]
Hi @citycide,

Yeah, I think I like the object and function syntaxes better than the array
one, just sort of on aesthetics.

Something I miss from Rails' ActiveRecord is the ability to do the 'has
many', 'belongs to', 'has and belongs to many', etc. type relationships...
I didn't see any nice way to do that kind of stuff in either trilogy or
knex... any clues appreciated!

For my current project, I wound up actually switching away from trilogy to
taffy, since the data wound up just being a lot more document-oriented than
tables oriented.  The limitation of having everything fit in memory isn't
that bad, because it's not a huge amount of data.

But I will definitely keep trilogy in mind for more relational stuff!

Cheers!

**Comment by citycide** [comment]
@csterritt 

> Something I miss from Rails' ActiveRecord is the ability to do the 'has
many', 'belongs to', 'has and belongs to many', etc. type relationships...
I didn't see any nice way to do that kind of stuff in either trilogy or
knex... any clues appreciated!

You're looking for [bookshelf](https://github.com/tgriesser/bookshelf), an ORM on top of knex by the same author.

I appreciate the feedback so far even if you haven't found trilogy useful for this specific project.

**Comment by csterritt** [comment]
Awesome!  Thanks!  I'll give bookshelf a try.  I knew there was probably
something out there :-)

**Comment by csterritt** [comment]
Hey -- sorry if I'm becoming the user from hell :-) but I can't figure out
how to get bookshelf to work with trilogy.

bookshelf wants to be initialized with a knex instance, and (from looking
at the trilogy source), 'db.queryBuilder' seems to bring back the knex
instance trilogy has built.  But it's failing for me... here's my test code.

Any help greatly appreciated!

    --chris s.

// import Trilogy from 'trilogy'
const Trilogy = require('trilogy')

// errorListener function
function errorHandler(err) {
  console.log(err.message)
}

// check if the file exists already
const dbPath = 'test.db'
const fs = require('fs')
let fileExists = fs.existsSync(dbPath)

// WITH OPTIONS:
// verbose function
const db = new Trilogy(dbPath, {
  verbose: console.log.bind(console),
  errorListener: errorHandler
})

if (!fileExists) {
  db.createTable('users', [
    { name: 'id', type: 'increments' },
    { name: 'name', type: 'string' },
    { name: 'age', type: 'integer' },
    { name: 'email', type: 'string' },
  ])
    .then(() => {
      db.insert('users', {
        name: 'Bob',
        age: 17,
        email: 'bob@bobbob.bob'
      })
  })
}

let bookshelf = require('bookshelf')(db.queryBuilder())

let User = bookshelf.Model.extend({
  tableName: 'users'
})

User.where('id', 1).fetch().then((user) => {
  console.log("User:", JSON.stringify(user))
})

**Comment by citycide** [comment]
I wouldn't recommend using bookshelf with trilogy. Bookshelf does a lot of relationship mapping using the connection with the sqlite3 ( or pg / mysql / etc ) driver. Trilogy is a just a layer between manually writing queries and passing them to sql.js There's no real connection going on.

**Comment by csterritt** [comment]
Thanks for the reply - I appreciate it.

**Comment by citycide** [comment]
This is solved in v0.11.1 which is available as the `@next` tag on npm and will eventually become v1.0.0.

```js
db.model('movies', { id: 'increments' }, { timestamps: true })
  .then(movies => { /* `movies` model is ready */ })
```

Thanks for all your input!

**Comment by csterritt** [comment]
Excellent, thanks.
